File: DigiBrainApp/app/src/main/java/com/dig/digibrain/services/server/ApiService.kt
package com.dig.digibrain.services.server

import com.dig.digibrain.models.*
import com.dig.digibrain.models.learnPaths.*
import com.dig.digibrain.models.postModels.quiz.AnswerPostModel
import com.dig.digibrain.models.postModels.quiz.QuestionPostModel
import com.dig.digibrain.models.postModels.quiz.QuizReportPostModel
import com.dig.digibrain.models.postModels.subject.ChapterPostModel
import com.dig.digibrain.models.postModels.subject.LessonPostModel
import com.dig.digibrain.models.postModels.subject.SubjectPostModel
import com.dig.digibrain.models.quiz.*
import com.dig.digibrain.models.subject.*
import okhttp3.MultipartBody
import retrofit2.Call
import retrofit2.http.*

interface ApiService {

    // Authentication
    @POST("api/Auth/login")
    suspend fun login(@Body model: LoginModel): TokenModel
    @POST("api/Auth/register")
    suspend fun register(@Body model: RegisterModel): UserModel
    @GET("api/Auth/object-storage-info")
    suspend fun getObjectStorageInfo(@Header("Authorization") authHeader : String): ObjectStorageInfoModel

    // User
    // --GET--
    @GET("api/users/{username}")
    suspend fun getUserDetailsByUsername(@Path("username") username: String): UserModel
    // --PUT--
    @PUT("api/users/{username}/classes/{classId}")
    suspend fun updateClass(
        @Header("Authorization") authHeader : String,
        @Path("username") username: String,
        @Path("classId") classId: Long): UserModel
    @Multipart
    @PUT("api/users/{username}/profile-image")
    suspend fun changeProfilePicture(
        @Path("username") username: String,
        @Part file: MultipartBody.Part): UserModel
    // --DELETE--
    @DELETE("api/users/{username}")
    suspend fun deleteAccount(@Header("Authorization") authHeader : String, @Path("username") username: String): UserModel

    // Language
    @GET("api/languages")
    suspend fun getLanguages(): List<LanguageModel>

    // Theory
    //--GET--
    @GET("api/classes/all")
    suspend fun getAllClasses(): List<ClassModel>
    @GET("api/domains")
    suspend fun getAllDomains(): List<DomainModel>
    @GET("api/classes/domains")
    suspend fun getDomainsForClass(
        @Query("Number") number: Int,
        @Query("AtUniversity") atUniversity: Boolean,
        @Query("LanguageId") languageId: Long
    ): List<DomainModel>
    @GET("api/domains/{id}")
    suspend fun getDomainById(@Path("id") id: Long): DomainModel
    @GET("api/classes")
    suspend fun getClassByNumberAndDomain(
        @Query("Number") number: Int,
        @Query("AtUniversity") atUniversity: Boolean,
        @Query("DomainId") DomainId: Long
    ): ClassModel
    @GET("/api/classes/{id}")
    suspend fun getClassById(@Path("id") id: Long): ClassModel
    @GET("api/subjects/class/{classId}/languages/{languageId}")
    suspend fun getSubjectsForClass(@Path("classId") classId: Long, @Path("languageId") languageId: Long): List<SubjectModel>
    @GET("api/chapters/subject/{subjectId}")
    suspend fun getChaptersForSubject(@Path("subjectId") subjectId: Long): List<ChapterModel>
    @GET("api/lessons/chapter/{chapterId}")
    suspend fun getLessonsForChapter(@Path("chapterId") chapterId: Long): List<LessonModel>
    //--POST--
    @POST("api/chapters")
    suspend fun addChapter(@Header("Authorization") authHeader : String, @Body model: ChapterPostModel): ChapterModel
    @POST("api/lessons")
    suspend fun addLesson(@Header("Authorization") authHeader : String, @Body model: LessonPostModel): LessonModel
    @POST("api/subjects")
    suspend fun addSubject(@Header("Authorization") authHeader : String, @Body model: SubjectPostModel): SubjectModel
    @POST("api/subjects/ids")
    suspend fun getSubjectsForIds(@Body subjectIds: List<Long>): List<SubjectModel>

    // Quiz
    // --GET--
    @GET("api/questions")
    suspend fun getRandomQuestions(
        @Query("Number") number: Int,
        @Query("Difficulty") difficulty: String,
        @Query("Type") type: String,
        @Query("LanguageId") languageId: Long
    ): List<QuestionModel>
    @GET("api/questions/subject/{id}")
    suspend fun getRandomQuestionsForSubject(
        @Path("id") subjectId: Long,
        @Query("Number") number: Int,
        @Query("Difficulty") difficulty: String,
        @Query("Type") type: String,
        @Query("LanguageId") languageId: Long
    ): List<QuestionModel>
    @GET("api/answers/questions/{questionId}")
    suspend fun getQuestionAnswers(@Path("questionId") questionId: Long): List<AnswerModel>
    @GET("api/reports/users/{username}")
    suspend fun getUserReports(@Path("username") username: String): List<QuizReportModel>
    // --POST--
    @POST("api/questions/ids")
    suspend fun getQuestionsForIds(@Body questionIds: List<Long>): List<QuestionModel>
    @POST("api/quizes/users/{username}")
    suspend fun createQuizForUser(@Path("username") username: String): QuizModel
    @POST("api/quizes")
    suspend fun addQuestionToQuiz(@Body model: QuestionAnswerModel): QuizQuestionModel
    @POST("api/questions")
    suspend fun createQuestion(@Header("Authorization") authHeader : String, @Body model: QuestionPostModel): QuestionModel
    @POST("api/answers/multiple")
    suspend fun addMultipleAnswers(@Header("Authorization") authHeader : String, @Body answers: List<AnswerPostModel>): List<AnswerModel>
    @POST("api/questions/{id}/subjects")
    suspend fun addQuestionToSubjects(@Header("Authorization") authHeader : String, @Path("id") questionId: Long, @Body subjectIds: List<Long>): List<SubjectQuestionModel>
    @POST("api/reports")
    suspend fun addUserReport(@Body model: QuizReportPostModel): QuizReportModel

    // Learn Path
    // --GET--
    @GET("api/learn-paths")
    suspend fun getLearnPaths(): List<LearnPathModel>
    @GET("api/learn-paths/{id}")
    suspend fun getLearnPathDetails(@Path("id") id: Long): LearnPathExpandedModel
    @GET("api/learn-paths/users/{username}/status")
    suspend fun getLearnPathsStatusForUsername(@Path("username") username: String): List<LearnPathStatusModel>
    // --POST--
    @POST("api/learn-paths/{id}/users/{username}/status")
    suspend fun startLearnPath(
        @Header("Authorization") authHeader : String,
        @Path("id") id: Long,
        @Path("username") username: String): LearnPathStatusModel
    // --PUT--
    @PUT("api/learn-paths/{id}/users/{username}/status")
    suspend fun updateLearnPathStatus(
        @Header("Authorization") authHeader : String,
        @Path("id") id: Long,
        @Path("username") username: String,
        @Body model: LearnPathStatusUpdateModel): LearnPathStatusModel
    @PUT("api/learn-paths/{id}/users/{username}/status/quiz")
    suspend fun updateLearnPathQuizStatus(
        @Header("Authorization") authHeader : String,
        @Path("id") id: Long,
        @Path("username") username: String,
        @Body model: LearnPathQuizStatusUpdateModel): LearnPathStatusModel
}

File: digibrain-web/src/services/ApiService.js
import axios from 'axios';

const API_BASE_URL = 'http://ec2-3-65-36-184.eu-central-1.compute.amazonaws.com';

const apiService = {

    // Authentication
    login: (username, password) => {
        let user = {
            username: username,
            password: password
        }

        return axios.post(`${API_BASE_URL}/api/Auth/login`, user);
    },
    getObjectStorageInfo: (config) => {
        return axios.get(`${API_BASE_URL}/api/Auth/object-storage-info`, config)
    },

    // Language
    getLanguages: () => {
        return axios.get(`${API_BASE_URL}/api/languages`)
    },

    // User
    getUserDetails: (username, config) => {
        return axios.get(`${API_BASE_URL}/api/users/${username}`, config)
    },

    // Learn
    // -- GET --
    getClassByNumberAndDomain: (classNumber, university, domainId) => {
        return axios.get(`${API_BASE_URL}/api/classes`, {
            params: {
                Number: classNumber,
                AtUniversity: university,
                DomainId: domainId
            }
        })
    },
    getSubjects: (classId, languageId) => {
        return axios.get(`${API_BASE_URL}/api/subjects/class/${classId}/languages/${languageId}`)
    },
    getChapters: (subjectId) => {
        return axios.get(`${API_BASE_URL}/api/chapters/subject/${subjectId}`)
    },
    getLessons: (chapterId) => {
        return axios.get(`${API_BASE_URL}/api/lessons/chapter/${chapterId}`)
    },
    // -- POST --
    addLesson: (title, text, chapterId, config) => {
        let lesson = {
            title: title,
            text: text,
            chapterId: chapterId
        }

        return axios.post(`${API_BASE_URL}/api/lessons`, lesson, config);
    },
    addChapter: (number, name, subjectId, config) => {
        let chapter = {
            number: number,
            name: name,
            subjectId: subjectId
        }

        return axios.post(`${API_BASE_URL}/api/chapters`, chapter, config);
    },
    addSubject: (subjectName, classId, languageId, iconId, config) => {
        let subject = {
            name: subjectName,
            classId: classId,
            languageId: languageId,
            iconId: iconId
        }

        return axios.post(`${API_BASE_URL}/api/subjects`, subject, config);
    },
    // -- DELETE --
    deleteLesson: (id, config) => {
        return axios.delete(`${API_BASE_URL}/api/lessons/${id}`, config);
    },
    deleteChapter: (id, config) => {
        return axios.delete(`${API_BASE_URL}/api/chapters/${id}`, config);
    },
    // -- PUT --
    updateChapter: (id, number, name, subjectId, config) => {
        let chapter = {
            number: number,
            name: name,
            subjectId: subjectId
        }

        return axios.put(`${API_BASE_URL}/api/chapters/${id}`, chapter, config);
    },
    updateLesson: (id, title, text, chapterId, config) => {
        let lesson = {
            title: title,
            text: text,
            chapterId: chapterId
        }

        return axios.put(`${API_BASE_URL}/api/lessons/${id}`, lesson, config);
    },

    // Quiz
    // -- GET --
    getQuestionsForSubject: (subjectId, numberOfQuestions, difficulty, questionType, languageId) => {
        let questionParams = {
            number: numberOfQuestions,
            difficulty: difficulty,
            type: questionType,
            languageId: languageId
        }

        return axios.get(`${API_BASE_URL}/api/questions/subject/${subjectId}/answers`, { params: questionParams })
    },
    getAnswersForQuestion: (questionId) => {
        return axios.get(`${API_BASE_URL}/api/answers/questions/${questionId}`)
    },
    addQuestion: (text, difficulty, questionType, languageId, config) => {
        let question = {
            text: text,
            difficulty: difficulty,
            type: questionType,
            languageId: languageId
        }

        return axios.post(`${API_BASE_URL}/api/questions`, question, config)
    },
    addAnswers: (answers, config) => {
        return axios.post(`${API_BASE_URL}/api/answers/multiple`, answers, config)
    },
    addQuestionToSubjects: (subjectIds, questionId, config) => {
        return axios.post(`${API_BASE_URL}/api/questions/${questionId}/subjects`, subjectIds, config)
    },
    // -- DELETE --
    deleteQuestion: (id, config) => {
        return axios.delete(`${API_BASE_URL}/api/questions/${id}`, config)
    },
    // -- PUT --
    updateQuestion: (id, text, difficulty, type, languageId, answers, config) => {
        let model = {
            text: text,
            difficulty: difficulty,
            type: type,
            languageId: languageId,
            answers: answers
        }

        return axios.put(`${API_BASE_URL}/api/questions/${id}`, model, config)
    },

    // Learn Paths
    // -- GET --
    getLearnPaths: () => {
        return axios.get(`${API_BASE_URL}/api/learn-paths`)
    },
    getLearnPathDetails: (id) => {
        return axios.get(`${API_BASE_URL}/api/learn-paths/${id}`)
    },

    // -- POST --
    addLearnPath: (title, description, author, subjectId, imageName, config) => {
        let learnPath = {
            title: title,
            description: description,
            author: author,
            subjectId: subjectId,
            imageName: imageName
        }

        return axios.post(`${API_BASE_URL}/api/learn-paths`, learnPath, config)
    },

    addLearnPathSection: (number, title, iconId, pathLearnId, config) => {
        let section = {
            number: number,
            title: title,
            iconId: iconId,
            pathLearnId: pathLearnId
        }

        return axios.post(`${API_BASE_URL}/api/learn-paths/sections`, section, config)
    },

    addLearnPathLesson: (number, title, description, pathSectionId, config) => {
        let lesson = {
            number: number,
            title: title,
            description: description,
            pathSectionId: pathSectionId
        }

        return axios.post(`${API_BASE_URL}/api/learn-paths/sections/lessons`, lesson, config)
    },

    addLearnPathTheory: (number, title, text, pathLessonId, config) => {
        let theory = {
            number: number,
            title: title,
            text: text,
            pathLessonId: pathLessonId
        }

        return axios.post(`${API_BASE_URL}/api/learn-paths/sections/lessons/theory`, theory, config)
    },

    addLearnPathQuiz: (questionsIds, pathLessonId, config) => {
        let quiz = []
        for(let i = 0; i < questionsIds.length; i++) {
            quiz.push( {
                score: 10,
                questionId: questionsIds[i],
                pathLessonId: pathLessonId
            })
        }

        return axios.post(`${API_BASE_URL}/api/learn-paths/sections/lessons/quiz`, quiz, config)
    },

    // -- DELETE --
    deleteLearnPathSection: (id, config) => {
        return axios.delete(`${API_BASE_URL}/api/learn-paths/sections/${id}`, config)
    },
    deleteLearnPathLesson: (id, config) => {
        return axios.delete(`${API_BASE_URL}/api/learn-paths/sections/lessons/${id}`, config)
    },
    deleteLearnPathTheory: (id, config) => {
        return axios.delete(`${API_BASE_URL}/api/learn-paths/sections/lessons/theory/${id}`, config)
    },
    deleteLearnPathQuiz: (id, config) => {
        return axios.delete(`${API_BASE_URL}/api/learn-paths/sections/lessons/${id}/quiz`, config)
    },

    // -- PUT --
    updateLearnPath: (id, title, description, author, subjectId, imageName, config) => {
        let learnPath = {
            title: title,
            description: description,
            author: author,
            subjectId: subjectId,
            imageName: imageName
        }

        return axios.put(`${API_BASE_URL}/api/learn-paths/${id}`, learnPath, config)
    },
    updateLearnPathSection: (id, number, title, iconId, pathLearnId, config) => {
        let section = {
            number: number,
            title: title,
            iconId: iconId,
            pathLearnId: pathLearnId
        }

        return axios.put(`${API_BASE_URL}/api/learn-paths/sections/${id}`, section, config)
    },
    updateLearnPathLesson: (id, number, title, description, pathSectionId, config) => {
        let lesson = {
            number: number,
            title: title,
            description: description,
            pathSectionId: pathSectionId
        }

        return axios.put(`${API_BASE_URL}/api/learn-paths/sections/lessons/${id}`, lesson, config)
    },
    updateLearnPathTheory: (id, number, title, text, pathLessonId, config) => {
        let theory = {
            number: number,
            title: title,
            text: text,
            pathLessonId: pathLessonId
        }

        return axios.put(`${API_BASE_URL}/api/learn-paths/sections/lessons/theory/${id}`, theory, config)
    },
    updateLearnPathQuiz: (id, questionsIds, pathLessonId, config) => {
        let quiz = []
        for(let i = 0; i < questionsIds.length; i++) {
            quiz.push( {
                score: 10,
                questionId: questionsIds[i],
                pathLessonId: pathLessonId
            })
        }

        return axios.put(`${API_BASE_URL}/api/learn-paths/sections/lessons/${id}/quiz`, quiz, config)
    },

    // Services
    notify: (title, message, topic, config) => {
        let configPost = {
            method: 'POST',
            headers: config.headers
        }

        return fetch(`${API_BASE_URL}/api/services/notify/title/${title}/message/${message}/topic/${topic}`, configPost)
    }
};

export default apiService;
